### ISSUE TITLE
some-bug-fixes

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/HelixLM#6]
# Potential Bugs in Trainer and Dataset

## To do:

Run ablations on the corrected code, verify that the tests on long and short text both pass and either improve performance or no regression in performance.

### Bug: last-batch gradient accumulation scaling

In `Trainer.train_epoch`:

```python
if self.grad_accum_steps > 1:
    loss = loss / self.grad_accum_steps   # always divides by 8
```

When the epoch ends with an incomplete buffer (e.g., only 3 gradients accumulated), you step the optimizer with a gradient that is **3× smaller than it should be** because you still divided by 8.

**Fix:** scale by the actual accumulated count on the trailing step.

```python
# In train_epoch, replace the unconditional division:
if self.grad_accum_steps > 1:
    loss = loss / self.grad_accum_steps

# With:
if self.grad_accum_steps > 1:
    # If this is the final, incomplete accumulation step, divide by actual count
    is_last = (batch_idx + 1) == len(self.train_loader)
    if is_last and accum_count < self.grad_accum_steps - 1:
        divisor = accum_count + 1   # +1 because we haven't incremented yet
    else:
        divisor = self.grad_accum_steps
    loss = loss / divisor
```
